worker/controller/manifold_test.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controller_test
+
+import (
+	"errors"
+	"time"
+
+	jujutesting "github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	worker "gopkg.in/juju/worker.v1"
+
+	coreagent "github.com/juju/juju/agent"
+	"github.com/juju/juju/state"
+	statetesting "github.com/juju/juju/state/testing"
+	coretesting "github.com/juju/juju/testing"
+	workercontroller "github.com/juju/juju/worker/controller"
+	"github.com/juju/juju/worker/dependency"
+	dt "github.com/juju/juju/worker/dependency/testing"
+)
+
+type ManifoldSuite struct {
+	statetesting.StateSuite
+	manifold             dependency.Manifold
+	openControllerCalled bool
+	openControllerErr    error
+	config               workercontroller.ManifoldConfig
+	agent                *mockAgent
+	resources            dt.StubResources
+}
+
+var _ = gc.Suite(&ManifoldSuite{})
+
+func (s *ManifoldSuite) SetUpTest(c *gc.C) {
+	s.StateSuite.SetUpTest(c)
+
+	s.openControllerCalled = false
+	s.openControllerErr = nil
+
+	s.config = workercontroller.ManifoldConfig{
+		AgentName:              "agent",
+		StateConfigWatcherName: "state-config-watcher",
+		OpenController:         s.fakeOpenController,
+		PingInterval:           10 * time.Millisecond,
+	}
+	s.manifold = workercontroller.Manifold(s.config)
+	s.resources = dt.StubResources{
+		"agent":                dt.StubResource{Output: new(mockAgent)},
+		"state-config-watcher": dt.StubResource{Output: true},
+	}
+}
+
+func (s *ManifoldSuite) fakeOpenController(coreagent.Config) (*state.Controller, error) {
+	s.openControllerCalled = true
+	if s.openControllerErr != nil {
+		return nil, s.openControllerErr
+	}
+	// Here's one we prepared earlier...
+	return s.Controller, nil
+}
+
+func (s *ManifoldSuite) TestInputs(c *gc.C) {
+	c.Assert(s.manifold.Inputs, jc.SameContents, []string{
+		"agent",
+		"state-config-watcher",
+	})
+}
+
+func (s *ManifoldSuite) TestStartAgentMissing(c *gc.C) {
+	s.resources["agent"] = dt.StubResource{Error: dependency.ErrMissing}
+	w, err := s.startManifold(c)
+	c.Check(w, gc.IsNil)
+	c.Check(err, gc.Equals, dependency.ErrMissing)
+}
+
+func (s *ManifoldSuite) TestStateConfigWatcherMissing(c *gc.C) {
+	s.resources["state-config-watcher"] = dt.StubResource{Error: dependency.ErrMissing}
+	w, err := s.startManifold(c)
+	c.Check(w, gc.IsNil)
+	c.Check(err, gc.Equals, dependency.ErrMissing)
+}
+
+func (s *ManifoldSuite) TestStartOpenControllerNil(c *gc.C) {
+	s.config.OpenController = nil
+	manifold := workercontroller.Manifold(s.config)
+	w, err := manifold.Start(s.resources.Context())
+	c.Check(w, gc.IsNil)
+	c.Check(err, gc.ErrorMatches, "OpenController is nil in config")
+}
+
+func (s *ManifoldSuite) TestStartNotController(c *gc.C) {
+	s.resources["state-config-watcher"] = dt.StubResource{Output: false}
+	w, err := s.startManifold(c)
+	c.Check(w, gc.IsNil)
+	c.Check(err, gc.Equals, dependency.ErrMissing)
```

Comment:
Does this mean that agents will try to start the thing, and then end up with an 'error not controller' in their logs, or the whole thing just won't start.

I'd like to avoid having failures-that-are-expected in log files.